egs/librispeech/asr/simple_v1/mmi_contextnet_decode.py

```diff
@@ -0,0 +1,274 @@
+    table = np.zeros((num_phones + 1, num_phones + 2))
+    table[:, 0] = 0
+    table[0, -1] = 0  # the start state has no arcs to the final state
+    assert P.arcs.dim0() == num_phones + 2
+    arcs = P.arcs.values()[:, :3]
+    probability = P.scores.exp().tolist()
+
+    assert arcs.shape[0] - num_phones == num_phones * (num_phones + 1)
+    for i, arc in enumerate(arcs.tolist()):
+        src_state, dest_state, label = arc[0], arc[1], arc[2]
+        prob = probability[i]
+        if label != -1:
+            assert label == dest_state
+        else:
+            assert dest_state == num_phones + 1
+        table[src_state][dest_state] = prob
+
+    try:
+        from prettytable import PrettyTable
+    except ImportError:
+        print('Please run `pip install prettytable`. Skip printing')
+        return
+
+    x = PrettyTable()
+
+    field_names = ['source']
+    field_names.append('sum')
+    for i in phone_ids:
+        field_names.append(phone_symbol_table[i])
+    field_names.append('final')
+
+    x.field_names = field_names
+
+    for row in range(num_phones + 1):
+        this_row = []
+        if row == 0:
+            this_row.append('start')
+        else:
+            this_row.append(phone_symbol_table[row])
+        this_row.append('{:.6f}'.format(table[row, 1:].sum()))
+        for col in range(1, num_phones + 2):
+            this_row.append('{:.6f}'.format(table[row, col]))
+        x.add_row(this_row)
+    with open(filename, 'w') as f:
+        f.write(str(x))
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument(
+        '--epoch',
+        type=int,
+        default=10,
+        help="Decoding epoch.")
+    parser.add_argument(
+        '--avg',
+        type=int,
+        default=5,
+        help="Number of checkpionts to average. Automaticly select "
+             "consecutive checkpoints before checkpoint specified by'--epoch'. ")
+    return parser
+
+
+def main():
+    parser = get_parser()
+    LibriSpeechAsrDataModule.add_arguments(parser)
+    args = parser.parse_args()
+
+    epoch = args.epoch
+    avg = args.avg
+
+    exp_dir = Path('exp-contextnet-noam-mmi-musan-sa')
+    setup_logger('{}/log/log-decode'.format(exp_dir), log_level='debug')
+
+    # load L, G, symbol_table
+    lang_dir = Path('data/lang_nosp')
+    symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
+    phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
+
+    phone_ids = get_phone_symbols(phone_symbol_table)
+    P = create_bigram_phone_lm(phone_ids)
+
+    phone_ids_with_blank = [0] + phone_ids
+    ctc_topo = k2.arc_sort(build_ctc_topo(phone_ids_with_blank))
+
+    logging.debug("About to load model")
+    # Note: Use "export CUDA_VISIBLE_DEVICES=N" to setup device id to N
+    # device = torch.device('cuda', 1)
+    device = torch.device('cuda')
+
+    model = ContextNet(
+        num_features=80,
```

Comment:
If you are using the default with 23 layers each with 5 blocks, I think this may be too big/deep for a relatively small amount of data.  I'd try 12 layers.

Reply:
With 12 blocks, we can get better results. 
23 blocks results are:
```
[test-clean] %WER 13.39% [7041 / 52576, 887 ins, 965 del, 5189 sub ]
[test-other] %WER 35.57% [18619 / 52343, 1327 ins, 3534 del, 13758 sub ]
```
12 blocks results are:
```
[test-clean] %WER 9.26% [4867 / 52576, 599 ins, 441 del, 3827 sub ]
[test-other] %WER 26.04% [13632 / 52343, 1382 ins, 1488 del, 10762 sub ]
```